_posts/2019-09-20-lambda-to-openfaas.md

```diff
@@ -0,0 +1,330 @@
+
+```js
+exports.handler = async (event) => {
+    // TODO implement
+    const response = {
+        statusCode: 200,
+        body: JSON.stringify('Hello from Lambda!'),
+    };
+    return response;
+};
+```
+
+Similarly, when you create a new function with the OpenFaaS CLI and the `node10-express` template, sample code is provided as shown here:
+```js
+"use strict"
+
+module.exports = (event, context) => {
+    let err;
+    const result =             {
+        status: "You said: " + JSON.stringify(event.body)
+    };
+    context
+        .status(200)
+        .succeed(result);
+}
+```
+
+At a quick glance, the two function samples are very similar. The main difference is in the invocation parameters. OpenFaaS passes in the `context` object which is used to provide additional context to the result. The context includes the properties `status`, `succeed`, `error`, and others that make discovering functionality and managing the response of the function easy without having to dig through pages of documentation.
+
+As you can see, migrating the sample function from Lambda to OpenFaaS would be as easy as adding the `context` object to the invocation, and moving the values from the inline `response` object to the OpenFaaS included `context` object properties.
+
+Next, we'll work through a real-world Lambda function that includes IAM permissions to access a DynamoDB table.
+
+## AWS Lambda
+
+The Lambda functions we'll be migrating are part of a URL shortener. One function creates and stores the shortened URL in DynamoDB, and the other function is responsible for looking up the destination URL, and redirecting the browser.
+
+Let's have a look at the `shortener` function which creates and stores the shortened URLs:
+
+```javascript
+const AWS = require('aws-sdk');
+
+const shortDomain = 'https://checkthisnow.net/'
+let docClient = new AWS.DynamoDB.DocumentClient();
+
+exports.handler = function (event, context, callback) {
+    let originalUrl = JSON.parse(event.body).url || '';
+
+    checkExists(originalUrl, function (exist) {
+        if (!exist) {
+            createAndStorePath(originalUrl, callback);
+        } else {
+            let shortenedURL = shortDomain + exist.token;
+            let response = {
+                isBase64Encoded: false,
+                statusCode: 200,
+                headers: {},
+                body: JSON.stringify({
+                    url: shortenedURL
+                })
+            };
+            callback(null, response);
+        }
+    });
+}
+
+function checkExists(url, callback) {
+    docClient.query(params, function (err, data) {
+        ...
+    });
+}
+
+function createAndStorePath(originalUrl, callback) {
```

Comment:
How are you authenticating clients that have write access to the API?

Reply:
I skipped the authentication mostly to simplify the post. I did add a sentence mentioning that the "shortener" creation function could be accessible in a private network to limit the access

Reply:
Please can you mention this in the blog post and suggest how you may use authentication? Leaving it open is OK for development, may be unwise to promote on the blog without giving context.

Reply:
Edit 1 ^